### PR TITLE
Safely navigate and compare nested hashes in the workflow feature

### PR DESCRIPTION
**Pull request by rubhanazeem**

### CODE REVIEW

**Comment by rubhanazeem** [comment]
[Safely navigate and compare nested hashes in the workflow feature](https://trello.com/c/Q9dn47bX/1739-safely-navigate-and-compare-nested-hashes-in-the-workflow-feature)

**Comment by dmarcoux** [comment]
The commit message is quite generic. How about this: `Safely navigate and compare nested hashes in the workflow feature`?

File: src/api/app/validators/workflow_filters_validator.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 class WorkflowFiltersValidator < ActiveModel::Validator
   def validate(record)
     @scm_extractor_payload = record.scm_extractor_payload
-    @workflow_instructions = record.workflow_instructions.with_indifferent_access
+    @workflow_instructions = record.workflow_instructions.deep_symbolize_keys
```
Comment (dmarcoux) [inline src/api/app/validators/workflow_filters_validator.rb:4]: `record` refers to an instance of `Workflow` and we already do `deep_symbolize_keys` on `workflow_instructions` in `Workflow.initalize`, so `deep_symbolize_keys` isn't needed here.

```suggestion
    @workflow_instructions = record.workflow_instructions
```

File: src/api/app/services/trigger_controller_service/scm_extractor.rb [hunk 2]
```diff
@@ -30,51 +30,51 @@ def call
...
+        pr_number: @payload[:number],
+        source_branch: @payload.dig(:pull_request, :head, :ref),
+        target_branch: @payload.dig(:pull_request, :base, :ref),
+        action: @payload[:action], # TODO: Names may differ, maybe we need to find our own naming (defer to service?)
+        source_repository_full_name: @payload.dig(:pull_request, :head, :repo, :full_name),
+        target_repository_full_name: @payload.dig(:pull_request, :base, :repo, :full_name),
         event: @event,
         api_endpoint: api_endpoint
-      }.with_indifferent_access
+      }.deep_symbolize_keys
```
Comment (dmarcoux) [inline src/api/app/services/trigger_controller_service/scm_extractor.rb:58]: All hash keys are already symbols here and keys of any nested hash possibly contained in `@payload` was already converted above in the `initialize` with `@payload = payload.deep_symbolize_keys`.

So we can safely remove `deep_symbolize_keys`:
```suggestion
      }
```

File: src/api/app/services/trigger_controller_service/scm_extractor.rb [hunk 3]
```diff
@@ -30,51 +30,51 @@ def call
...
+        pr_number: @payload.dig(:object_attributes, :iid),
+        source_branch: @payload.dig(:object_attributes, :source_branch),
+        target_branch: @payload.dig(:object_attributes, :target_branch),
+        action: @payload.dig(:object_attributes, :action), # TODO: Names may differ, maybe we need to find our own naming (defer to service?)
+        project_id: @payload.dig(:object_attributes, :source_project_id),
+        path_with_namespace: @payload.dig(:project, :path_with_namespace),
         event: @event,
         api_endpoint: "#{uri.scheme}://#{uri.host}"
-      }.with_indifferent_access
+      }.deep_symbolize_keys
```
Comment (dmarcoux) [inline src/api/app/services/trigger_controller_service/scm_extractor.rb:77]: Same comment as for `github_extractor_payload`.

Here, we can also safely remove `deep_symbolize_keys`:
```suggestion
      }
```

**Review by dmarcoux — DISMISSED** [review]

File: src/api/spec/models/workflow_spec.rb [hunk 4]
```diff
@@ -200,7 +200,7 @@
           scm: 'github',
           action: 'opened',
           event: 'pull_request'
-        }.with_indifferent_access
+        }.deep_symbolize_keys
```
[the remark this answers is not in the document]
Comment (rubhanazeem) [reply src/api/spec/models/workflow_spec.rb]: Done.

File: src/api/spec/support/shared_contexts/a_scm_payload_hash.rb [hunk 5]
```diff
@@ -10,7 +10,7 @@
       action: 'opened',
       repository_full_name: 'openSUSE/open-build-service',
       event: 'pull_request'
-    }.with_indifferent_access
+    }.deep_symbolize_keys
```
[the remark this answers is not in the document]
Comment (rubhanazeem) [reply src/api/spec/support/shared_contexts/a_scm_payload_hash.rb]: Done.

File: src/api/spec/support/shared_contexts/a_scm_payload_hash.rb [hunk 6]
```diff
@@ -25,6 +25,6 @@
       project_id: 1,
       path_with_namespace: 'gitlabhq/gitlab-test',
       event: 'Merge Request Hook'
-    }.with_indifferent_access
+    }.deep_symbolize_keys
```
[the remark this answers is not in the document]
Comment (rubhanazeem) [reply src/api/spec/support/shared_contexts/a_scm_payload_hash.rb]: Done.

File: src/api/app/services/trigger_controller_service/scm_extractor.rb (hunk 2 above)
Comment (rubhanazeem) [reply src/api/app/services/trigger_controller_service/scm_extractor.rb]: Done.

File: src/api/app/models/workflow.rb [hunk 7]
```diff
@@ -13,10 +13,10 @@ class Workflow
 
   attr_accessor :workflow_instructions, :scm_extractor_payload, :token
 
-  # Overwriting the initializer is needed to set `with_indifferent_access`
+  # Overwriting the initializer is needed to set `deep_symbolize_keys`
```
Comment (dmarcoux) [inline src/api/app/models/workflow.rb:16]: This comment isn't really useful... it's pretty obvious. I would remove it.

File: src/api/app/models/workflow.rb [hunk 8]
```diff
@@ -13,10 +13,10 @@ class Workflow
...
   def initialize(workflow_instructions:, scm_extractor_payload:, token:)
```
Comment (dmarcoux) [inline src/api/app/models/workflow.rb:17]: We should stick to the conventions of `initialize` for an `ActiveModel::Model` (see the [code block](https://api.rubyonrails.org/classes/ActiveModel/Model.html) with `def initialize(attributes={})`). The small advantage is not having to set `@token`. It's not a blocker for this PR, but I think we should do it while we're here...
```ruby
def initialize(attributes = {})
    super
    @workflow_instructions = attributes[:workflow_instructions].deep_symbolize_keys
    @scm_extractor_payload = attributes[:scm_extractor_payload].deep_symbolize_keys
end
```

File: src/api/app/jobs/report_to_scm_job.rb [hunk 9]
```diff
@@ -35,7 +35,7 @@ def matched_event_subscription
     EventSubscriptionsFinder.new
                             .for_scm_channel_with_token(event_type: @event_type, event_package: @event_package)
                             .select do |event_subscription|
-      Workflows::Filter.new(filters: event_subscription.payload.with_indifferent_access[:workflow_filters]).match?(@event)
+      Workflows::Filter.new(filters: event_subscription.payload.deep_symbolize_keys[:workflow_filters]).match?(@event)
```
[the remark this answers is not in the document]
Comment (rubhanazeem) [reply src/api/app/jobs/report_to_scm_job.rb:38]: Okay, I've updated `Workflows::Filter`.

File: src/api/app/models/workflow.rb (hunk 8 above)
Comment (rubhanazeem) [reply src/api/app/models/workflow.rb]: Done. It's updated.

File: src/api/app/services/workflows/filter.rb [hunk 10]
```diff
@@ -2,12 +2,12 @@ module Workflows
...
       filters ||= {}
-      @repository_filters = filters[:repositories]
-      @architecture_filters = filters[:architectures]
+      @repository_filters = filters[:repositories].deep_symbolize_keys
+      @architecture_filters = filters[:architectures].deep_symbolize_keys
     end
 
     def match?(event)
-      match_repository?(event.payload['repository']) && match_architecture?(event.payload['arch'])
+      match_repository?(event.payload.deep_symbolize_keys[:repository]) && match_architecture?(event.payload.deep_symbolize_keys[:arch])
```
Comment (dmarcoux) [inline src/api/app/services/workflows/filter.rb:10]: Why not do `deep_symbolize_keys` once and store the result?

File: src/api/app/services/workflows/filter.rb [hunk 11]
```diff
@@ -2,12 +2,12 @@ module Workflows
   class Filter
     def initialize(filters:)
       filters ||= {}
-      @repository_filters = filters[:repositories]
-      @architecture_filters = filters[:architectures]
+      @repository_filters = filters[:repositories].deep_symbolize_keys
```
Comment (dmarcoux) [inline src/api/app/services/workflows/filter.rb:5]: Doing `deep_symbolize_keys` once directly on `filters` would be better.

**Review by danidoni — DISMISSED** [review]

File: src/api/app/services/workflows/filter.rb (hunk 10 above)
Comment (rubhanazeem) [reply src/api/app/services/workflows/filter.rb]: As discussed, I'm staying with strings where values are coming from DB.

File: src/api/app/services/workflows/filter.rb (hunk 11 above)
Comment (rubhanazeem) [reply src/api/app/services/workflows/filter.rb]: Done

**Pull request merged**